### DISCUSSION TITLE
Possible to get list of states, enumerate variables, get results?

### DISCUSSION THREAD

**Discussion by bilderbuchi** [OpenModelica/OMSimulator#1128] [Q&A] [answered]
I am trying to find a periodic steady state for a model, i.e. where the solution is fully periodic in time, see [here](https://github.com/OpenModelica/OpenModelica/discussions/8609).

As available OM optimisation facilities are not appropriate for such a problem, I'm currently trying to attack this (scripted in Python) by 
* simulating an FMU of the model for one period and 
* varying parameters defining the intial values of all relevant variables (IIUC, the _states_)
* until these values agree with the values at the end of one period

I am wondering if it is possible/feasible in OMSimulator to :
1. query for FMU information like a list of variables
2. critically: get  a list of all _state_ variables (I want to check that they all have appropriate start values, resp. adjust those until convergence)
3. for convenience: get the simulation result not as a file, but as a data structure (e.g. for plotting)

Or is this not something OMSimulator is intended to provide, and I should rather turn towards PyFMI of FMPy? 
I've already found ['get_states_list()](https://jmodelica.org/pyfmi/pyfmi.html#pyfmi.fmi.FMUModelBase2.get_states_list) which seems to be what I need.

**Comment by arun3688** [comment, marked as answer]
@bilderbuchi we are planning to implement such features and it will be available quite soon.

**Reply by lochel** [reply]
@bilderbuchi The following example shows how to query a list of all variables with type and kind attributes.

```python
import OMSimulator as oms
import xml.etree.ElementTree as ET
import pprint
[...]

model = oms.newModel("myModel")
[...]

# extract all variables with type and kind attributes
_signals = {}
signalFilter = model.exportSnapshot(':resources/signalFilter.xml')
root = ET.fromstring(signalFilter)
for var in root[0][0]:
  name = var.attrib['name']
  type_ = var.attrib['type']
  kind = var.attrib['kind']
  _signals[name] = {'type': type_, 'kind': kind}
pprint.pprint(_signals, width=120)
```

And this is an example output produced with the code above:
```
{'equationPair.root.equation1.der(_D_outputAlias_x2)': {'kind': 'unknown', 'type': 'Real'},
 'equationPair.root.equation1.x1':                     {'kind': 'input', 'type': 'Real'},
 'equationPair.root.equation1.x2':                     {'kind': 'output', 'type': 'Real'},
 'equationPair.root.equation2.der(_D_outputAlias_x1)': {'kind': 'unknown', 'type': 'Real'},
 'equationPair.root.equation2.x1':                     {'kind': 'output', 'type': 'Real'},
 'equationPair.root.equation2.x2':                     {'kind': 'input', 'type': 'Real'}}
```

Everything else you requested will be available soon as mentioned by @arun3688 😃

**Reply by bilderbuchi** [reply]
That helps, thanks! 
Great to hear -- I didn't see an open issue, so wasn't aware that it's on your radar. :-)

**Reply by bilderbuchi** [reply]
@lochel that was useful, although it does not seem to be possible to identify states based e.g. on the `kind` attribute.
For now I'm parsing the name for the substring `.der(`, based on the assumption that whatever follows in the parenthesis after `der`, i.e. everything that has a derivative, is a state. 
Is that a 100% valid assumption? I'd say I'm 90% confident 😁

**Reply by lochel** [reply]
Right, this approach also came to our mind but it is unfortunately not sufficient. Even though it will work in most cases I suppose. Another idea was to put “state” instead for “unknown” into the signalFilter. But that is also not sufficient since, e.g., a variable could be output and state at the same time.
One way to solve this is to expose the entire modelDescription file and do the parsing in python. Or to provide a specific api call to query exactly what is needed, e.g. the list of states. I prefer the second of the two ideas. What is your opinion?
